vald/evm/evm.go

```diff
@@ -66,116 +66,121 @@ func (mgr Mgr) ProcessNewChain(event *types.ChainAdded) (err error) {
-				events = append(events, evmTypes.Event{
-					Chain: event.Chain,
-					TxID:  event.TxID,
-					Index: uint64(i),
-					Event: &evmTypes.Event_Transfer{
-						Transfer: &erc20Event,
-					},
-				})
-			default:
-			}
+	txReceipt, err := mgr.getTxReceiptIfFinalized(event.Chain, common.Hash(event.TxID), event.ConfirmationHeight)
+	if err != nil {
+		return err
+	}
+	if txReceipt == nil {
+		mgr.logger.Info(fmt.Sprintf("broadcasting empty vote for poll %s", event.PollID.String()))
+		_, err := mgr.broadcaster.Broadcast(context.TODO(), voteTypes.NewVoteRequest(mgr.cliCtx.FromAddress, event.PollID, types.NewVoteEvents(event.Chain, nil)))
+
+		return err
+	}
+
+	var events []types.Event
+	for i, log := range txReceipt.Logs {
+		if log.Topics[0] != ERC20TransferSig {
+			continue
+		}
 
+		if !bytes.Equal(event.TokenAddress.Bytes(), log.Address.Bytes()) {
+			continue
 		}
-		return true
-	})
 
-	msg := voteTypes.NewVoteRequest(mgr.cliCtx.FromAddress, event.PollID, evmTypes.NewVoteEvents(event.Chain, events))
+		erc20Event, err := decodeERC20TransferEvent(log)
+		if err != nil {
+			mgr.logger.Debug(sdkerrors.Wrap(err, "decode event Transfer failed").Error())
+			continue
+		}
+
+		if erc20Event.To != event.DepositAddress {
+			continue
+		}
+
+		if err := erc20Event.ValidateBasic(); err != nil {
+			mgr.logger.Debug(sdkerrors.Wrap(err, "invalid event Transfer").Error())
+			continue
+		}
+
+		events = append(events, types.Event{
+			Chain: event.Chain,
+			TxID:  event.TxID,
+			Index: uint64(i),
+			Event: &types.Event_Transfer{
+				Transfer: &erc20Event,
+			},
+		})
+	}
+
 	mgr.logger.Info(fmt.Sprintf("broadcasting vote %v for poll %s", events, event.PollID.String()))
-	_, err := mgr.broadcaster.Broadcast(context.TODO(), msg)
+	_, err = mgr.broadcaster.Broadcast(context.TODO(), voteTypes.NewVoteRequest(mgr.cliCtx.FromAddress, event.PollID, types.NewVoteEvents(event.Chain, events)))
+
 	return err
 }
 
 // ProcessTokenConfirmation votes on the correctness of an EVM chain token deployment
-func (mgr Mgr) ProcessTokenConfirmation(event *evmTypes.ConfirmTokenStarted) error {
+func (mgr Mgr) ProcessTokenConfirmation(event *types.ConfirmTokenStarted) error {
 	if !slices.Any(event.Participants, func(v sdk.ValAddress) bool { return v.Equals(mgr.validator) }) {
 		mgr.logger.Debug("ignoring token confirmation poll: not a participant", "pollID", event.PollID)
 		return nil
 	}
 
-	rpcClient, found := mgr.rpcs[strings.ToLower(event.Chain.String())]
-	if !found {
-		return fmt.Errorf("unable to find an RPC for chain '%s'", event.Chain)
+	txReceipt, err := mgr.getTxReceiptIfFinalized(event.Chain, common.Hash(event.TxID), event.ConfirmationHeight)
+	if err != nil {
+		return err
 	}
+	if txReceipt == nil {
+		mgr.logger.Info(fmt.Sprintf("broadcasting empty vote for poll %s", event.PollID.String()))
+		_, err := mgr.broadcaster.Broadcast(context.TODO(), voteTypes.NewVoteRequest(mgr.cliCtx.FromAddress, event.PollID, types.NewVoteEvents(event.Chain, nil)))
```

Comment:
nil events will break the falsy result check
https://github.com/axelarnetwork/axelar-core/blob/667a5702f2cb8c24c9d9892445cdcd4cd746dc11/x/evm/keeper/vote_handler.go#L47

Reply:
No it does not since `len(nil)` is 0. I also tested it to make sure.

Reply:
good

Reply:
I can also convert the constructor to be like `NewVoteEvents(chain nexus.ChainName, events ...Event)` if that makes it look better.